### ISSUE TITLE
[QUESTION] Full websocket support

### ISSUE THREAD

**Issue by perrosnk** [hasura/graphql-engine#4644]
Is it possible to use Hasura with full websocket support (connecting with websockets for queries, mutations, subscriptions)?

**Comment by shahidhk** [comment]
@perrosnk Yes, Hasura supports executing queries and mutations over websockets apart from subscriptions.

**Comment by perrosnk** [comment]
@shahidhk So, how does authentication work with websockets? When using a webhook does the token get validated at the beginning of the connection or is it validated with each operation?
I couldn't find any document about it.

**Comment by shahidhk** [comment]
@perrosnk The authentication happens with `connection_init` when the websocket is connected and is valid until the expiry of the JWT (if using JWT auth mode) or as per the `Expires` header set on the webhook response.

Once authenticated, all operations are allowed without further check, until the authentication expires.

Tagging @marionschleifer to convert this to docs.

**Comment by perrosnk** [comment]
1. Does the websocket accept connection from anonymous users if I return x-hasura-role = anonymous from the webhook?

2. What do you exactly mean "all operations are allowed"? All operations according to the user-role? Are the hasura permissions getting validated against the user-role?

**Comment by shahidhk** [comment]
> Does the websocket accept connection from anonymous users if I return x-hasura-role = anonymous from the webhook?

Yes, it will, and the role will be set to `anonymous`.

> What do you exactly mean "all operations are allowed"? All operations according to the user-role? Are the hasura permissions getting validated against the user-role?

I meant whatever is allowed per that user role and session variables will be allowed.

Executing operations over HTTP and WS are no different except that one is a long-living connection, so make sure you set up ways to expire the connection with the authentication expiring.

**Comment by osaxma** [comment]
Hi @shahidhk , 

I've two related questions, please:

### 1- How to reject all anonymous connection attempts to the websocket endpoint? 

Even though I don't have `HASURA_GRAPHQL_UNAUTHORIZED_ROLE` set to anything, and the engine is in JWT authentication mode, I still can connect to the websocket anonymously. The connection doesn't get rejected and stays alive (I keep receiving  "{"type":"ka"}" message but the queries are rejected. I would like the server to terminate the websocket connection if there's no authentication header. Is that possible?

### 2 - How to update authentication of an active connection after token expiry? 
Currently, I'm keeping track of all subscriptions, and when the token expires _(once per hour for firebase auth)_, I call `terminate_connection` and then reconnect again with the new token and restart all  subscriptions that i'm keeping track of.... 

Is there a way where I can send the updated token only without having to go through stopping all subscriptions and then restart them again?

Thank you and I appreciate any leads.
